src/test/java/org/opensearch/security/transport/SecurityInterceptorTests.java

```diff
@@ -201,17 +292,306 @@ public <T extends TransportResponse> void sendRequest(
+        OpenSearchSecurityPlugin.GuiceHolder guiceHolder = new OpenSearchSecurityPlugin.GuiceHolder(
+            mock(RepositoriesService.class),
+            transportService,
+            mock(IndicesService.class),
+            mock(PitService.class),
+            mock(ExtensionsManager.class)
+        );
+        // CS-ENFORCE-SINGLE
+
+        DiscoveryNode localNode = new DiscoveryNode("local-node1", new TransportAddress(localAddress, 1234), Version.CURRENT);
+        Connection connection1 = transportService.getConnection(localNode);
+
+        DiscoveryNode remoteNode = new DiscoveryNode("remote-node", new TransportAddress(localAddress, 6789), remoteNodeVersion);
+        Connection connection2 = transportService.getConnection(remoteNode);
+
+        // checking thread context inside sendRequestDecorate for connection1 with null local node -- we should serialize this
+        AsyncSender sender = new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                String serializedUserHeader = threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+                assertEquals(serializedUserHeader, Base64Helper.serializeObject(user, useJDKSerialization));
+            }
+        };
+
+        // this is a request where the local node is null; have to use the remote connection since the serialization will fail
+        securityInterceptor.sendRequestDecorate(sender, connection2, action, request, options, handler, null);
+
+        // from original context
+        User transientUser1 = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        assertEquals(transientUser1, user);
+        assertEquals(threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER), null);
+
+        // test sending various missing decorates with missing request info
+
+        // checking thread context inside sendRequestDecorate
+        sender = new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                User transientUser = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+                assertEquals(transientUser, user);
+            }
+        };
+
+        // Having null connection will cause useJDKSerialization to throw npe
+        AsyncSender finalSender = sender; // Have to do this because of being a runnable
```

Comment:
Seems like you are creating a sender, then calling sendRequestDecorate(...) many times in this same method, seems like this single test could be refactored into multiple tests where it is clearer what is being verified so there is a cleaner pattern setup/act/assert.

What do you think?

Reply:
Yeah good idea.